### PR TITLE
Add orgs

### PR DESCRIPTION
**Pull request by mattkrick**
Things left to do:

 @ackernaut here are the things i'm stuck on

Designs:
- [x] Org field for `/newteam`
 - Field to select an existing org
 - Field (or link) to type in a new org that expands to add billing info
- [x] Organizations-list: Create new org
 - modal or link to create a new org and add billing
- [x] Organization-detail: Add new billing leader
 - Pop a modal/dropdown box with a search box & name of all org members
- [x] Organization-detail: Billing info empty state
 - Design what the billing info section should look like if no billing info exists
 - Design the modal asking for CC and expiry
- [x] Organization-detail: Update billing info
 - The modal that pops up when you click update under billing info
- [x] Organization-detail: Users
 - See: [Visual concept on Dropbox](https://www.dropbox.com/s/mvkm653vod620rs/me-organization-members%402x.png?dl=0)
- [x] Organization-detail: Invoice See Details
 - See: #652
- [x] Organization-detail: edit picture
 - allow user to upload from desktop or type in a url for their new picture

Matt's to-do:
- [x] wire up changing org avatar
- [x] support account balance in the invoices
- [x] cancel subscriptions when payments get rejected (requires changing settings in stripe)
- [x] validate `isPaid === true` before subscribing to any subscription except notifications (but disregard most/all mutations because we'd rather be performant for the 99% than be slow & stingy towards the 1%)
- [x] create org dropdown menu for /newteam
- [x] write addOrg mutation
- [x] implement orgMembers list view
- [x] implement vertical 3 dot menu for orgMembersRow
- [x] handle invitation approval success mutation
- [x] handle invitation rejection mutation (and the front-end modal)
- [x] pop modal when team belongs to an unpaid org
- [x] show a "!" on the sidebar next to unpaid teams
- [x] show a bell on the sidebar linking to notifications
- [x] disable multiple clicks for step 2 of the welcome wizard
- [x] store activeUsers instead of activeUserCount so we can reactively subscribe to orgs 
- [x] wire up the front end of invoices
- [x] figure out what to do about current invoices-- calculate on the fly? 
- [x] create a whitelist of notification types & copy those over to new billing leaders when promoted. removed those notification when demoted. 
- [x] add a "pending org leader approval" group to the team settings so team leaders can understand why some folks don't join.

### CODE REVIEW

File: src/server/database/migrations/20161125141100-addOrgs.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    await Promise.all(indices);
+  } catch (e) {
+  }
+
+  const waitIndices = [
+    r.table('Team').indexWait('orgId'),
+    r.table('Notification').indexWait('orgId', 'parentId', 'userId'),
+    r.table('User').indexWait('email', 'orgs', 'billingLeaderOrgs')
+  ];
+  await Promise.all(waitIndices);
```
Comment (jordanh) [inline src/server/database/migrations/20161125141100-addOrgs.js]: Nice! I was wondering when we were going to need to use this. And, here it is!

File: src/server/socketHandlers/scConnectionHandler.js [hunk 2]
```diff
@@ -43,15 +43,39 @@ export default function scConnectionHandler(exchange) {
...
-    const tmsIsValid = isTmsValid(tms, authToken.tms);
+    const tmsDB = await r.table('User').get(userId)('tms');
+    const tmsIsValid = isTmsValid(tmsDB, tms);
     if (timeLeftOnToken < REFRESH_JWT_AFTER || !tmsIsValid) {
-      authToken.tms = tms;
+      authToken.tms = tmsDB;
       socket.setAuthToken(authToken);
     }
+    // no need to wait for this, it's just for billing
+    r.branch(
```
Comment (jordanh) [inline src/server/socketHandlers/scConnectionHandler.js]: This query is so rad.

File: src/universal/components/Action/Action.js [hunk 3]
```diff
@@ -1,13 +1,13 @@
 import React, {PropTypes} from 'react';
 import withStyles from 'universal/styles/withStyles';
 import {css} from 'aphrodite-local-styles/no-important';
-import Notifications from 'universal/modules/notifications/containers/Notifications/Notifications';
+import Toast from 'universal/modules/toast/containers/Toast/Toast';
```
Comment (jordanh) [inline src/universal/components/Action/Action.js]: Bless your heart for renaming this.

File: src/universal/containers/Signout/signout.js [hunk 4]
```diff
@@ -2,7 +2,7 @@ import {cashay} from 'cashay';
 import {removeAuthToken} from 'universal/redux/authDuck';
 import {reset as resetAppState} from 'universal/redux/rootDuck';
 import {segmentEventTrack} from 'universal/redux/segmentActions';
-import {showSuccess} from 'universal/modules/notifications/ducks/notifications';
+import {showSuccess} from 'universal/modules/toast/ducks/toastDuck';
```
Comment (jordanh) [inline src/universal/containers/Signout/signout.js]: toastDuck sounds delicious

**Comment by jordanh** [comment]
Wow man. Crazy amount of work in here.

Incidentally, I love how you refactored (user|team)DashboardContainer into a single container. Nice work!

**Comment by ackernaut** [comment]
@mattkrick updated these:

- Organization-detail: Users
- Organization-detail: Invoice See Details

Going to work on some UI options for updating org avatars.

**Comment by mattkrick** [comment]
thinking aloud here:
- orgs:users is M:F (many-to-few)
- 3NF would say we have Org 1->M OrgMembers F<-1 User, but if we made a new table, then we'd have to subscribe to that OrgUser table & either denormalize user name, email, picture onto the table, or trigger a lookup on each changefeed.  
- we need to subscribe to all users in an org (/organizations/:orgId/members) including their status & role
- we need to subscribe to all orgs that a user has access to (/newteam, /organizations)
- we need a cheap way to determine active/inactive users on each organizations (ie we can't touch the user table to get it)
- we need the role (eg billingAdmin) and the status (active, inactive) for each user

since we need to subscribe to these things, an intermediary table doesn't make sense.
since a user only has a few orgs, it makes sense to store org data on the user doc instead of user data on the org doc.

- [ ] can reactively sub to all orgs that a user is a part of (`r.table('Organization').getAll(userId, {index: 'userIds'})`)
- [ ] can reactively sub to all orgs that a user is a leader of `r.table('Organization').getAll(userId, {index: 'ownerIds'})`)
- [ ] can reactively sub to all users (+ role + status) that belong to an org (`r.table('User').getAll(orgId...)`)

Then second one can be achieved by using the same index as the first, but then filtering by the billingLeaderOrgs on the client. Or, we could change userIds to users & have an array of docs. each doc could store the role & active status. Then activeUserCount is users filtered by status, etc. If we do this, then on the last one we'll need to keep the orgs array on the user doc. To figure out their role & status, we'd need to join that with the data from the org. But, we don't always need that, only when they click Members, which I assume will be infrequent. So, we'll need a second subscription just for the orgUsers & then have to do an ugly join on the client. Alternatively, we could store role on the user doc as well, and inactive is already there on the user doc. 

current:
org = {activeUsers, inactiveUsers}
users = {inactive, orgs, billingLeaderOrgs}
fails because: no way to determine if a user has admin access to an org without looking on the user doc. That means we have to send down all the orgs they're a part of, send down their billingLeaderOrgs field, and then perform a join on the client. 

proposed:
org = {users: {id, role, inactive}}
users = {org: {id, role}, inactive}

this lets us do everything with only 1 arbitrary multi index on each table.

**Comment by jordanh** [comment]
@mattkrick, this makes good sense. If following your thinking is a good razor...

**Comment by ackernaut** [comment]
Thanks for the explanation @mattkrick, seems safe to me! Glad you guys grok this better than I do. :)

**Comment by jordanh** [comment]
# Test Plan

(this is going to grow a lot as I interact with the new feature set)

## From New Team Leader

### It:

- [ ] should allow you to create a new new user and
- [ ] should allow you to complete the welcome wizard
- [ ] should give you a 30 day trial
- [ ] should notify you that you can add another 30 days (if TODO variable is rewritten in the DB)

**Comment by mattkrick** [comment]
end-to-end tests for the PR:
it
 - billing leader can create a new team in their own org
   - expect them to get updated tms token
   - expect invitations to be sent out to invitees
 - user can create a new team in their own org
   - expect them to get updated tms token
   - expect invitations to be sent out to org members
   - expect notifications to be sent out to billing leaders
   - expect pending approvals to be added for non-members
   - expect invitations and pending approvals to be visible on the team settings
   - expect to cancel a pending approval
   - expect notification to show up on billing leaders user dash
   - expect notification to have an accept and reject option
   - expect accept option to remove the pending org approval and notification and send an invite
   - expect reject option to pop a modal asking why
   - expect that reason or a default is turned into a notification visible to the requesting user
   - expect that rejecteduser notification to be removed via OK
 - user can create their own org
   - expect new team form turns into new org form allowing for org name
   - expect stripeToken is captured via CC modal
   - expect CC modal has an error state if CC info is wrong
   - expect org is created
   - expect orgUsers has the userId of the creator in it
   - expect userOrgs has the new org appended
   - expect new tms token is sent down to client
   - expect invitation sent out to all invitees
   - expect new org and sub is created in stripe
   - expect quantity to be set to 1
 - invited users can join org
 - org leaders can pause users
   - expect inactive to be on the user doc
   - expect inactive to be in the orgUsers docs for all the orgs they're in
 - logging in (sc connection) can unpause
   - expect to update lastSeenAt
   - expect to unset inactive state from user & orgUsers
   - expect the stripe quantity to ++ FOR EACH ORG
   - expect to receive 2*orgs invoice item lines from stripe
   - create an invoiceitemhook 
 - autopause kicks in after 30 days
   - action-chronos sends a message to pause inactives every day
   - expect `pauseInactiveUsers` to pause all users that haven't logged on in 30 days
   - expect to set inactive state for each user and all orgUsers
   - expect stripe quantity--
   - expect to receive 2*orgs invoice lines from stripe
   - 

it 
 - attribute invoice lines to users state changes
   - write invoiceitemhook for each...
 - ends the subscription when a trial ends without conversion
   - expect teams to have a ! next to their names
   - expect a modal to be on the team dash with a link to the org billing if they're a billing leader
   - expect a modal to be on team with telling them their billing leaders suck
   - expect updating the credit card to create a new subscription w/ o a trial period and with the current quantity
   - expect teams to get reactivated after successful payment 
   - expect billing leader to have a notification bar at the top of dashboards
   - 
 - ends the subscription when a payment is rejected  
 - handles an incoming payment invoice from stripe
   - expect that line to transform into a line showing the date range the user's email was paused
   - expect the start date to be the beginning of the last invoice or the last time the user went on pause (max)

File: package.json [hunk 5]
```diff
@@ -15,7 +15,7 @@
   "main": "index.js",
   "scripts": {
     "start": "NODE_ENV=production node ./src/server/server.babel.js",
-    "dev": "NODE_ENV=development node ./src/server/server.babel.js",
+    "dev": "if [ ! -f \"dll/vendors.json\" ] ; then npm run build:dll ; fi ; NODE_ENV=development node ./src/server/server.babel.js",
```
Comment (mattkrick) [inline package.json]: your command of SHELL never ceases to amaze me...

**Comment by mattkrick** [comment]
more tests

- [ ] handles team member leaving team
   - [ ] TeamMember is set to isNotRemoved: false
   - [ ] user's tms is updated
   - [ ] auth0 tms is updated
   - [ ] new token is pushed to the kicked out user
   - [ ] kick out message is sent to client
   - [ ] server kicks them out of all subs that pertain to the team (presence, team, team members, etc)
   - [ ] if user is on the team dash, or any child route like /teamId/settings, they are redirected

**Comment by mattkrick** [comment]
remaining tests
 - [ ] creates an invoice for the user
   - [ ] expect a message from stripe saying the invoice is being created
   - [ ] expect to fetch all invoice line items
   - [ ] expect a line item for next months charges (assuming quantity stays the same)
   - [ ] expect detailed line item for each added user including date they were added & prorated cost
   - [ ] expected detailed line item for each removed user including date removed & prorated cost
   - [ ] expect detailed line item for each period of inactivity with a start date of the max(time they were paused, invoice start) and an end date of the time they were unpaused, with a prorated cost
   - [ ] expect aggregated line items for all users added, removed, and paused
   - [ ] expect line item for current balance, if applicable
   - [ ] expect total charge to be next month's charge + this months adjustments + balance
 - [ ] shows you a previous invoice 
   - [ ] expect a list of previous invoices to be shown in the org billing (where !pending)
   - [ ] expect clicking on an invoice to take you to a printer-friendly invoice
 - [ ] shows you the current invoice
   - [ ] expect the invoice amount to come directly from stripe
   - [ ] expect clicking on the invoice to return something that looks similar to other invoices
   - [ ] expect to reuse the current pending invoice if it was cached less than an hour ago
   - [ ] expect to create a new pending invoice if it was last cached more than an hour ago
- [ ] handles users on 0 teams but 1 org
   - [ ] expect to be able to go to new team
 - [ ] can invite ex-teammember ex-org members
   - [ ] expect user to get an email inviting them back into the org
   - [ ] expect accepting that invite will reactivate their teamMember doc
   - [ ] expect their user doc tms to be correct
   - [ ] expect their TeamMember doc to be activate
   - [ ] expect their user auth token to include the teamId
 - [ ] can remove a user from an org
   - [x] expect isNotRemoved to be false on every TeamMember doc for that userOrg
   - [x] expect tms to be reduced on their user doc
   - [x] expect if they were team lead to promote the next-oldest as team lead
   - [x] expect the team to be deleted if they were the only one on it
   - [x] expect all projects to be reassigned to the team lead
   - [x] expect all team actions to be marked as completed
   - [x] expect a new authToken is sent down to that user with the reduced tms
   - [x] expect them to get kicked out of team things if they are currently online (with a toast)
   - [x] expect user to be removed from the org
   - [x] expect stripe quantity--
   - [x] expect userOrgs to have the org removed
   - [x] expect orgUsers to have the user removed
   - [x] expect invoiceitemhook to be created
   - [x] expect 2 invoice lines to come from stripe
   - [x] expect to update those 2 invoice lines with a REMOVE_USER type and the userId
   - [ ] expect a "leave" modal if the person to remove is yourself
   - [ ] expect no option if the person is yourself & you are the only billing leader
   - [ ] expect any org-related notification to go away 
 - [ ] can remove a user from a team & is now no longer part of the org
   - [ ] expect all the things that happen when a team member is removed from a team (above)
   - [ ] expect user to be removed from the org
   - [ ] expect stripe quantity--
   - [ ] expect userOrgs to have the org removed
   - [ ] expect orgUsers to have the user removed
   - [ ] expect invoiceitemhook to be created
   - [ ] expect 2 invoice lines to come from stripe
   - [ ] expect to update those 2 invoice lines with a REMOVE_USER type and the userId
 - [ ] shows a CTA if i don't own any orgs
   - [ ] if no orgs are present, the /organizations route shows a CTA to create my own org
 - [ ] starts a new subscription after payment was rejected
   - [ ] expect a new subscription to be started in stripe with the correct quantity and id added to org
   - [ ] expect trial ended notification (or payment rejected) to go away
   - [ ] expect teams to be marked as paid
   - [ ] expected updated periodStart, periodEnd

**Comment by mattkrick** [comment]
invoice todos:
- store unit price on the invoice itself
- break the `lines` array down by type
- store nextPeriodStart/end on invoice

**Pull request merged**